File: macos/ServerPolicy.swift
import Foundation
import CryptoKit

enum FileMCPPolicyProfiles {
    static let restricted = "restricted"
    static let workspaceAuto = "workspace-auto"
    static let custom = "custom"
    static let legacyCommandCompatible = "legacy-command-compatible"

    static func isKnown(_ value: String) -> Bool {
        [restricted, workspaceAuto, custom, legacyCommandCompatible].contains(value)
    }

    static func isUserSelectable(_ value: String) -> Bool {
        [restricted, workspaceAuto, custom].contains(value)
    }
}

struct LocalPolicyConfiguration {
    var profile: String = FileMCPPolicyProfiles.restricted
    var customMaxRisk: String = "low"
    var customAllowedEffects: [String] = ["read", "metadata"]
    var customAllowNetworkOpenWorld = false
    var customAllowShell = false

    static func fromLegacy(enableCommands: Bool) -> LocalPolicyConfiguration {
        LocalPolicyConfiguration(
            profile: enableCommands ? FileMCPPolicyProfiles.legacyCommandCompatible : FileMCPPolicyProfiles.restricted
        )
    }

    func normalized() throws -> LocalPolicyConfiguration {
        let normalizedProfile = profile.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard FileMCPPolicyProfiles.isKnown(normalizedProfile) else {
            throw ToolCatalogError.invalid("Unsupported policy profile: \(profile)")
        }
        let normalizedRisk = customMaxRisk.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard ["low", "medium", "high"].contains(normalizedRisk) else {
            throw ToolCatalogError.invalid("Unsupported custom policy risk: \(customMaxRisk)")
        }
        let validEffects: Set<String> = ["read", "write", "delete", "execute", "external", "metadata"]
        let effects = Array(Set(customAllowedEffects
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
            .filter { !$0.isEmpty })).sorted()
        guard effects.allSatisfy({ validEffects.contains($0) }) else {
            throw ToolCatalogError.invalid("Custom policy contains an unsupported effect")
        }
        return LocalPolicyConfiguration(
            profile: normalizedProfile,
            customMaxRisk: normalizedRisk,
            customAllowedEffects: effects,
            customAllowNetworkOpenWorld: customAllowNetworkOpenWorld,
            customAllowShell: customAllowShell
        )
    }
}

struct PolicySnapshot: Equatable {
    let generation: UInt64
    let hash: String
}

final class ServerPolicy {
    private struct State {
        let generation: UInt64
        let hash: String
        let configuration: LocalPolicyConfiguration
    }

    private let lock = NSLock()
    private var state: State

    init(configuration: LocalPolicyConfiguration) throws {
        let normalized = try configuration.normalized()
        state = State(generation: 1, hash: try Self.computeHash(normalized), configuration: normalized)
    }

    static func fromLegacy(enableCommands: Bool) throws -> ServerPolicy {
        try ServerPolicy(configuration: .fromLegacy(enableCommands: enableCommands))
    }

    var profile: String {
        lock.lock(); defer { lock.unlock() }
        return state.configuration.profile
    }

    var generation: UInt64 {
        lock.lock(); defer { lock.unlock() }
        return state.generation
    }

    var hash: String {
        lock.lock(); defer { lock.unlock() }
        return state.hash
    }

    var legacyUnsafeGitCompatibility: Bool {
        lock.lock(); defer { lock.unlock() }
        return state.configuration.profile == FileMCPPolicyProfiles.legacyCommandCompatible
    }

    func capture() -> PolicySnapshot {
        lock.lock(); defer { lock.unlock() }
        return PolicySnapshot(generation: state.generation, hash: state.hash)
    }

    func update(_ configuration: LocalPolicyConfiguration) throws {
        let normalized = try configuration.normalized()
        let nextHash = try Self.computeHash(normalized)
        lock.lock()
        defer { lock.unlock() }
        guard nextHash != state.hash else { return }
        state = State(generation: state.generation &+ 1, hash: nextHash, configuration: normalized)
    }

    func isAllowed(_ toolName: String) -> Bool {
        guard let metadata = try? CanonicalToolCatalog.shared.toolPolicyMetadata(named: toolName) else { return false }
        lock.lock(); defer { lock.unlock() }
        return Self.isAllowed(state.configuration, metadata: metadata)
    }

    func filterDefinitions(_ definitions: [[String: Any]]) throws -> [[String: Any]] {
        try definitions.compactMap { definition in
            guard let name = definition["name"] as? String else {
                throw ToolCatalogError.invalid("Invalid tool definition while applying policy")
            }
            return isAllowed(name) ? definition : nil
        }
    }

    func authorize(_ toolName: String, prepared: PolicySnapshot? = nil) throws {
        let metadata = try CanonicalToolCatalog.shared.toolPolicyMetadata(named: toolName)
        lock.lock()
        defer { lock.unlock() }
        if let prepared,
           prepared.generation != state.generation || prepared.hash != state.hash {
            throw ToolCatalogError.invalid("Prepared operation is stale because local policy changed")
        }
        guard Self.isAllowed(state.configuration, metadata: metadata) else {
            throw ToolCatalogError.invalid("Tool is denied by active local policy: \(toolName)")
        }
    }

    func metadata() -> [String: Any] {
        lock.lock(); defer { lock.unlock() }
        return [
            "profile": state.configuration.profile,
            "generation": NSNumber(value: state.generation),
            "hash": state.hash,
        ]
    }

    private static func isAllowed(_ configuration: LocalPolicyConfiguration, metadata: ToolPolicyMetadata) -> Bool {
        switch configuration.profile {
        case FileMCPPolicyProfiles.restricted:
            return !metadata.capabilities.contains("process.shell") &&
                !metadata.capabilities.contains("process.exec")
        case FileMCPPolicyProfiles.legacyCommandCompatible:
            return true
        case FileMCPPolicyProfiles.workspaceAuto:
            return !metadata.capabilities.contains("process.shell")
                && !metadata.capabilities.contains("network.open_world")
                && metadata.effect != "external"
        case FileMCPPolicyProfiles.custom:
            guard riskRank(metadata.risk) <= riskRank(configuration.customMaxRisk) else { return false }
            guard configuration.customAllowedEffects.contains(metadata.effect) else { return false }
            if !configuration.customAllowNetworkOpenWorld && metadata.capabilities.contains("network.open_world") { return false }
            if !configuration.customAllowShell && metadata.capabilities.contains("process.shell") { return false }
            return true
        default:
            return false
        }
    }

    private static func riskRank(_ value: String) -> Int {
        switch value {
        case "low": return 0
        case "medium": return 1
        case "high": return 2
        default: return Int.max
        }
    }

    private static func computeHash(_ configuration: LocalPolicyConfiguration) throws -> String {
        let canonical = [
            configuration.profile,
            configuration.customMaxRisk,
            configuration.customAllowedEffects.joined(separator: ","),
            configuration.customAllowNetworkOpenWorld ? "true" : "false",
            configuration.customAllowShell ? "true" : "false",
        ].joined(separator: "\n")
        return SHA256.hash(data: Data(canonical.utf8)).map { String(format: "%02x", $0) }.joined()
    }
}

File: windows/src/FileMCP.Core/ServerPolicy.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace FileMCP.Core;

public static class FileMcpPolicyProfiles
{
    public const string Restricted = "restricted";
    public const string WorkspaceAuto = "workspace-auto";
    public const string Custom = "custom";
    public const string LegacyCommandCompatible = "legacy-command-compatible";

    public static bool IsKnown(string value) => value is Restricted or WorkspaceAuto or Custom or LegacyCommandCompatible;
    public static bool IsUserSelectable(string value) => value is Restricted or WorkspaceAuto or Custom;
}

public sealed class LocalPolicyConfiguration
{
    public string Profile { get; set; } = FileMcpPolicyProfiles.Restricted;
    public string CustomMaxRisk { get; set; } = "low";
    public List<string> CustomAllowedEffects { get; set; } = ["read", "metadata"];
    public bool CustomAllowNetworkOpenWorld { get; set; }
    public bool CustomAllowShell { get; set; }

    public static LocalPolicyConfiguration FromLegacy(bool enableCommands) => new()
    {
        Profile = enableCommands ? FileMcpPolicyProfiles.LegacyCommandCompatible : FileMcpPolicyProfiles.Restricted,
    };

    public LocalPolicyConfiguration CloneNormalized()
    {
        var profile = string.IsNullOrWhiteSpace(Profile) ? FileMcpPolicyProfiles.Restricted : Profile.Trim().ToLowerInvariant();
        if (!FileMcpPolicyProfiles.IsKnown(profile)) throw new FileMcpException($"Unsupported policy profile: {Profile}");
        var maxRisk = string.IsNullOrWhiteSpace(CustomMaxRisk) ? "low" : CustomMaxRisk.Trim().ToLowerInvariant();
        if (maxRisk is not ("low" or "medium" or "high")) throw new FileMcpException($"Unsupported custom policy risk: {CustomMaxRisk}");
        var effects = (CustomAllowedEffects ?? [])
            .Where(effect => !string.IsNullOrWhiteSpace(effect))
            .Select(effect => effect.Trim().ToLowerInvariant())
            .Distinct(StringComparer.Ordinal)
            .Order(StringComparer.Ordinal)
            .ToList();
        var validEffects = new HashSet<string>(new[] { "read", "write", "delete", "execute", "external", "metadata" }, StringComparer.Ordinal);
        if (effects.Any(effect => !validEffects.Contains(effect)))
            throw new FileMcpException("Custom policy contains an unsupported effect");
        return new LocalPolicyConfiguration
        {
            Profile = profile,
            CustomMaxRisk = maxRisk,
            CustomAllowedEffects = effects,
            CustomAllowNetworkOpenWorld = CustomAllowNetworkOpenWorld,
            CustomAllowShell = CustomAllowShell,
        };
    }
}

internal readonly record struct PolicySnapshot(long Generation, string Hash);

internal sealed class ServerPolicy
{
    private sealed record PolicyState(long Generation, string Hash, LocalPolicyConfiguration Configuration);
    private readonly object _gate = new();
    private PolicyState _state;

    public ServerPolicy(LocalPolicyConfiguration configuration)
    {
        var normalized = configuration.CloneNormalized();
        _state = new PolicyState(1, ComputeHash(normalized), normalized);
    }

    public static ServerPolicy FromLegacy(bool enableCommands) => new(LocalPolicyConfiguration.FromLegacy(enableCommands));

    public string Profile { get { lock (_gate) return _state.Configuration.Profile; } }
    public long Generation { get { lock (_gate) return _state.Generation; } }
    public string Hash { get { lock (_gate) return _state.Hash; } }
    public bool LegacyUnsafeGitCompatibility { get { lock (_gate) return _state.Configuration.Profile == FileMcpPolicyProfiles.LegacyCommandCompatible; } }

    public PolicySnapshot Capture()
    {
        lock (_gate) return new PolicySnapshot(_state.Generation, _state.Hash);
    }

    public void Update(LocalPolicyConfiguration configuration)
    {
        var normalized = configuration.CloneNormalized();
        var hash = ComputeHash(normalized);
        lock (_gate)
        {
            if (string.Equals(hash, _state.Hash, StringComparison.Ordinal)) return;
            _state = new PolicyState(checked(_state.Generation + 1), hash, normalized);
        }
    }

    public bool IsAllowed(string toolName)
    {
        var metadata = CanonicalToolCatalog.ToolPolicyMetadata(toolName);
        lock (_gate) return IsAllowed(_state.Configuration, metadata);
    }

    public JsonArray FilterDefinitions(JsonArray definitions)
    {
        var result = new JsonArray();
        foreach (var node in definitions)
        {
            if (node is not JsonObject definition || definition["name"] is not JsonValue value || !value.TryGetValue<string>(out var name))
                throw new FileMcpException("Invalid tool definition while applying policy");
            if (IsAllowed(name)) result.Add(definition.DeepClone());
        }
        return result;
    }

    public void Authorize(string toolName, PolicySnapshot? prepared = null)
    {
        var metadata = CanonicalToolCatalog.ToolPolicyMetadata(toolName);
        lock (_gate)
        {
            if (prepared.HasValue && (prepared.Value.Generation != _state.Generation || !string.Equals(prepared.Value.Hash, _state.Hash, StringComparison.Ordinal)))
                throw new FileMcpException("Prepared operation is stale because local policy changed");
            if (!IsAllowed(_state.Configuration, metadata))
                throw new FileMcpException($"Tool is denied by active local policy: {toolName}");
        }
    }

    public JsonObject Metadata()
    {
        lock (_gate)
        {
            return new JsonObject
            {
                ["profile"] = _state.Configuration.Profile,
                ["generation"] = _state.Generation,
                ["hash"] = _state.Hash,
            };
        }
    }

    private static bool IsAllowed(LocalPolicyConfiguration configuration, ToolPolicyMetadata metadata)
    {
        switch (configuration.Profile)
        {
            case FileMcpPolicyProfiles.Restricted:
                return !metadata.Capabilities.Contains("process.shell", StringComparer.Ordinal) &&
                       !metadata.Capabilities.Contains("process.exec", StringComparer.Ordinal);
            case FileMcpPolicyProfiles.LegacyCommandCompatible:
                return true;
            case FileMcpPolicyProfiles.WorkspaceAuto:
                return !metadata.Capabilities.Contains("process.shell", StringComparer.Ordinal) &&
                       !metadata.Capabilities.Contains("network.open_world", StringComparer.Ordinal) &&
                       metadata.Effect != "external";
            case FileMcpPolicyProfiles.Custom:
                if (RiskRank(metadata.Risk) > RiskRank(configuration.CustomMaxRisk)) return false;
                if (!configuration.CustomAllowedEffects.Contains(metadata.Effect, StringComparer.Ordinal)) return false;
                if (!configuration.CustomAllowNetworkOpenWorld && metadata.Capabilities.Contains("network.open_world", StringComparer.Ordinal)) return false;
                if (!configuration.CustomAllowShell && metadata.Capabilities.Contains("process.shell", StringComparer.Ordinal)) return false;
                return true;
            default:
                return false;
        }
    }

    private static int RiskRank(string value) => value switch { "low" => 0, "medium" => 1, "high" => 2, _ => int.MaxValue };

    private static string ComputeHash(LocalPolicyConfiguration configuration)
    {
        var canonical = string.Join("\n", new[]
        {
            configuration.Profile,
            configuration.CustomMaxRisk,
            string.Join(",", configuration.CustomAllowedEffects),
            configuration.CustomAllowNetworkOpenWorld ? "true" : "false",
            configuration.CustomAllowShell ? "true" : "false",
        });
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(canonical))).ToLowerInvariant();
    }
}
